django/db/models/fields/composite.py

```diff
@@ -0,0 +1,415 @@
+import copy
+from collections import OrderedDict
+from itertools import chain
+import re
+import weakref
+
+from django.core import checks
+from django.core.exceptions import FieldDoesNotExist
+from django.db.models.fields import Field
+from django.db.models.fields.composite_lookups import (
+    CompositeExact, CompositeIn, CompositeIsNull, SubfieldTransform
+)
+from django.utils import six
+from django.utils.functional import cached_property
+from django.utils.observables import Observable, ObservableDict
+
+
+__all__ = ('constrain', 'CompositeField')
+
+
+def constrain(*fields, **kwargs):
```

Comment:
If work is continued on that PR, `db_index` should be changed on all fields to accept an `Index` object or `None`, (with `True` defaulting to a BTree index), so that wouldn't change what is described here except for changing the accepted argument types to `db_index` . That PR also doesn't mention constraints -- the arguments to `constrain` can be extended to provide `primary_key=True` or `CHECK` constraints (if a suitable API is provided for them). I've limited myself to implementing `UNIQUE` constraints, because implementing other types of constraints is outside the scope of this PR.

I was hoping that `constrain` would deprecate `index_together` and `unique_together`, but I haven't been explicit about that. I'm more interested in standalone composite fields, but some people expressed an interest in (something like) this syntax for providing multi-column constraints for fields which already existed on the table and I didn't think it was a bad idea...